feat(compose): Introduce `kraft cloud compose` commands

craciunoiuc opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [ ] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

GitHub-Closes: #1426

craciunoiuc committed:
fix(create): Do not pretty print when asked to print as list

This is not correct, but in the long run I think we should add
a 'pretty' output type to differentiate.

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
fix(down): Fix incorrect struct name and underlying operation

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose start`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose ls`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose ps`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose create`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
refactor(down): Re-work to fit style and also remove service groups

This commit reworks the cloud instance down subcommand such that it:

1. Has passable `Auth` and `Client` non-attributes;
2. Use more generic KraftCloud client, making it more reusable;
3. Removes unnecessary error checks;
4. Removes lingering service groups;
5. Moves all necessary functionality into `Remove` method.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose build`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose push`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(logs): Make colorful printer public

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(compose): Introduce `kraft cloud compose logs`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

antoineco commented on internal/cli/kraft/cloud/compose/build/build.go:
```diff
@@ -0,0 +1,192 @@
+)
+
+type BuildOptions struct {
+	Auth   *config.AuthConfig           `noattribute:"true"`
+	Client kcinstances.InstancesService `noattribute:"true"`
+	Metro  string                       `noattribute:"true"`
+	Token  string                       `noattribute:"true"`
+
+	project     *compose.Project
+	composeFile string
```
(Comment applies to all commands)

Since those aren't attributes which can set directly by the user via command-line flags, do they need to be inside this struct at all?

I'm excluding `metro` and `token`, which are set in `Pre` and therefore need to be carried to `Run`.
https://github.com/unikraft/kraftkit/blob/1f3487e1c216deb6e966a67da33baaadca2d4c91/internal/cli/kraft/cloud/tunnel/tunnel.go#L26-L29

craciunoiuc replied:
hmm, makes sense, I just added them there because I saw them similar to `metro/token` + it made copy-paste easier

I'll move them down as variables

craciunoiuc replied:
fixed as new commit

craciunoiuc committed:
fix(pkg): Do not overwrite working directory if empty args

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
fix(utils): Always prepend workdir if rootfs not absolute

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

craciunoiuc committed:
feat(deploy): Introduce optional 'Image' name flag to deploy

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung committed:
feat(compose): Introduce `kraft cloud compose stop`

Signed-off-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

nderjung committed:
feat(utils): Introduce utility method for re-wrapping KC images

Additionally adjust both `kraft pkg` and `kraft build` such that
invocations with the "kraftcloud" platform are automatically
redirected to an alternative kraftcloud registry repository.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(build): Introduce new option `--no-rootfs`

This prevents construction of the rootfs.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(build): Return well-known error `ErrContextNotBuildable`

This helps catching the output from invoking `Build`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(build): Only populate workdir if option is not set

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(logs): Expose `Logs` as a general internal CLI method

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(logs): Handle piping logs from multiple instances

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(compose): Expose flag to set alternative Composefile path

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(stop): Re-work to be in the style and format of other cmds

This commit reworks the cloud instance stop subcommand such that it:

1. Has passable `Auth` and `Client` non-attributes;
2. Use more generic KraftCloud client, making it more reusable;
3. Removes unnecessary error checks;
4. Removes unused `Output` attribute;
5. Moves all necessary functionality into `Stop` method;
6. Uses lower-case log messages.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(remove): Re-work to be in the style and format of other cmds

This commit reworks the cloud service remove subcommand such that it:

1. Has passable `Auth` and `Client` non-attributes;
2. Use more generic KraftCloud client, making it more accessible;
3. Removes unnecessary error checks;
4. Moves all necessary functionality into `Remove` method;
5. Uses lower-case log messages.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(remove): Re-work to be in the style and format of other cmds

This commit reworks the cloud instance remove subcommand such that it:

1. Has passable `Auth` and `Client` non-attributes;
2. Use more generic KraftCloud client, making it more accessible;
3. Removes unnecessary error checks;
4. Moves all necessary functionality into `Remove` method;
5. Uses lower-case log messages.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(compose): Use `WithField` to display `Composefile`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
gomod(deps): Bump sdk.kraft.cloud from 0.5.7 to 0.5.8

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(up): Re-work `kraft cloud compose up`

This refactor adjusts the subcommand such that:

1. It performs the build, packaging and pushing of services
   based on the `build` context;
2. Maps networks to service groups;
3. Tails the logs if `--detach` is not specified.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
refactor(start): Re-work to be in the style and format of other cmds

This commit reworks the cloud instance start subcommand such that it:

1. Has passable `Auth` and `Client` non-attributes;
2. Use more generic KraftCloud client, making it more accessible;
3. Removes unnecessary error checks;
4. Uses lower-case log messages;
5. Overall code-path reduction.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc commented on internal/cli/kraft/cloud/compose/ps/ps.go:
```diff
@@ -0,0 +1,142 @@
+		Short:   "List the services in a KraftCloud deployment",
+		Use:     "ps [FLAGS]",
+		Args:    cobra.NoArgs,
+		Aliases: []string{"str"},
```
typo

nderjung committed:
feat(remove): Introduce `--wait-empty` which polls until empty

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(utils): Handle printing error message in table

Signed-off-by: Alexander Jung <alex@unikraft.io>
